### PR TITLE
Add type specifying extension for ctype_digit

### PR DESCRIPTION
**Pull request by fluffycondor**

### CODE REVIEW

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+			return new SpecifiedTypes();
+		}
+		if ($context->null()) {
+			throw new ShouldNotHappenException();
+		}
+
+		$argType = $scope->getType($node->getArgs()[0]->value);
+		$intRange = new UnionType([
+			IntegerRangeType::fromInterval(48, 57), // ASCII-codes for 0-9
+			IntegerRangeType::createAllGreaterThanOrEqualTo(256) // Starting from 256 ints are interpreted as strings
```
Comment (herndlm) [inline src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]:
```suggestion
			IntegerRangeType::createAllGreaterThanOrEqualTo(256), // Starting from 256 ints are interpreted as strings
```
the trailing comma is missing, as also reported by the failing linting test

Comment (fluffycondor) [reply src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]: Thanks!

**Comment by staabm** [comment]
You may find additional inspiration for the tests in https://github.com/phpstan/phpstan-src/pull/1615

**Comment by fluffycondor** [comment]
Thanks for those tests. I found some edge cases.

**Comment by BackEndTea** [comment]
Isnt passing ints to ctype_* functions being deprecated?

**Comment by staabm** [comment]
Yep it is. Supporting existing code which uses it is still good though to e.g. prevent false positives

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			throw new ShouldNotHappenException();
+		}
+
+		$types = [
+			IntegerRangeType::fromInterval(48, 57), // ASCII-codes for 0-9
+			IntegerRangeType::createAllGreaterThanOrEqualTo(256), // Starting from 256 ints are interpreted as strings
+		];
+
+		if ($context->true()) {
+			$types[] = new AccessoryNumericStringType();
```
Comment (ondrejmirtes) [inline src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]: numeric strings always have to be in combination with a basic type. So this should be IntersectionType of StringType and AccessoryNumericStringType.

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		}
+
+		return $this->typeSpecifier->create($node->getArgs()[0]->value, new UnionType($types), $context, false, $scope);
```
Comment (ondrejmirtes) [inline src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]: Call TypeCombinator::union() instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php (hunk 2 above)
Comment (fluffycondor) [reply src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]: Done

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php (hunk 3 above)
Comment (fluffycondor) [reply src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]: Done

**Comment by ondrejmirtes** [comment]
Please rebase the branch and solve conflicts, thanks :)

**Comment by ondrejmirtes** [comment]
No merge commits please, linear history only.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!
